src/application/machines/devtoolsMachine.tsx

```diff
@@ -1,143 +1,233 @@
-    },
-    retrying: {
-      on: {
-        connect: "connected",
-        clientNotFound: "notFound",
-      },
-      entry: ["notifyWaitingForConnection", "closeModal"],
-      after: {
-        connectTimeout: {
-          target: "notFound",
+        initializing: {
+          entry: "renderUI",
+          always: [
+            {
+              guard: "contextValid",
+              target: "ok",
+            },
+            {
+              target: "error",
+            },
+          ],
         },
-      },
-    },
-    connected: {
-      on: {
-        disconnect: "disconnected",
-      },
-      entry: ["notifyConnected", "closeModal"],
-      exit: ["resetStore"],
-      after: {
-        2500: {
-          actions: "closeBanner",
+        error: {
+          always: [
+            {
+              guard: "contextValid",
+              target: "ok",
+            },
+          ],
+          on: {},
+        },
+        ok: {
+          always: {
+            guard: not("contextValid"),
+            target: "error",
+          },
+          on: {},
         },
       },
     },
-    disconnected: {
+    connection: {
+      initial: "disconnected",
       on: {
-        connect: "connected",
-        timeout: "timedout",
-        clientNotFound: "notFound",
-      },
-      entry: "notifyDisconnected",
-      after: {
-        connectTimeout: {
-          target: "notFound",
+        "client.register": {
+          actions: [
+            assign({
+              registeredClients: ({ context }) => context.registeredClients + 1,
+            }),
+          ],
+        },
+        "client.terminated": {
+          actions: [
+            assign({
+              registeredClients: ({ context }) => context.registeredClients - 1,
+            }),
+          ],
+        },
+        "client.setCount": {
+          actions: [
+            assign({
+              registeredClients: ({ event }) => event.count,
+            }),
+          ],
         },
       },
-    },
-    timedout: {
-      entry: "notifyTimedOut",
-    },
-    notFound: {
-      on: {
-        retry: "retrying",
-        connect: "connected",
+      states: {
+        connected: {
+          entry: ["notifyConnected"],
+          exit: ["resetStore"],
+          after: {
+            2500: {
+              actions: "closeBanner",
+            },
+          },
+          always: {
+            guard: ({ context }) => context.registeredClients === 0,
+            target: "disconnected",
+          },
+        },
+        disconnected: {
+          entry: "notifyDisconnected",
+          invoke: {
+            id: "reconnect",
+            src: "reconnect",
+          },
+          always: {
+            guard: ({ context }) => context.registeredClients > 0,
+            target: "connected",
+          },
+        },
       },
-      entry: ["notifyNotFound", "openModal"],
     },
   },
 });
+
+export const DevToolsMachineContext = createContext<DevToolsActor | null>(null);
```

Comment:
We now use Context to pass this down into the React app - but not the `createContext` from xState, as that would create the actor from the machine and we want to create the actor outside of React (the Actor renders React).